prov/verbs/src/windows/verbs_nd_ibv.c

```diff
@@ -0,0 +1,871 @@
+			errno = H2F(hr);
+			*bad_wr = wr;
+			return errno;
+		}
+
+		wr = wr->next;
+	}
+
+	return FI_SUCCESS;
+}
+
+int ibv_post_send(struct ibv_qp *qp, struct ibv_send_wr *wr,
+		  struct ibv_send_wr **bad_wr)
+{
+	trace(__func__);
+	if ((!qp) || (!wr) || (!bad_wr)) {
+		errno = EINVAL;
+		return errno;
+	}
+
+	struct ibv_qp_vrb_nd *qp_nd =
+		container_of(qp, struct ibv_qp_vrb_nd, qp);
+
+	*bad_wr = NULL;
+	while (wr) {
+		uint32_t length = 0;
+		for (int i = 0; i < wr->num_sge; ++i) {
+			length += wr->sg_list[i].length;
+		}
+		uint32_t flags = 0;
+		if (length <= qp_nd->max_inline_data) {
+			flags |= ND_OP_FLAG_INLINE;
+		}
+		HRESULT hr = qp_nd->nd2qp->lpVtbl->Send(qp_nd->nd2qp,
+							(void *)wr->wr_id,
+							(ND2_SGE *)wr->sg_list,
+							wr->num_sge, flags);
+		VRB_DBG(FI_LOG_EP_DATA,
+			"IND2QueuePair::Send: hr=0x%08lx, nsge=%d, addr=0x%016llx, length=%d, lkey=0x%08lx\n",
+			hr, wr->num_sge, wr->sg_list[0].addr,
+			wr->sg_list[0].length, wr->sg_list[0].lkey);
+		if (FAILED(hr)) {
+			errno = H2F(hr);
+			*bad_wr = wr;
+			return errno;
+		}
+
+		wr = wr->next;
+	}
+
+	return FI_SUCCESS;
+}
+
+const char *ibv_wc_status_str(enum ibv_wc_status status)
+{
+	trace(__func__);
+	static const char *const wc_status_str[] = {
+		[IBV_WC_SUCCESS] = "success",
+		[IBV_WC_LOC_LEN_ERR] = "local length error",
+		[IBV_WC_LOC_QP_OP_ERR] = "local QP operation error",
+		[IBV_WC_LOC_EEC_OP_ERR] = "local EE context operation error",
+		[IBV_WC_LOC_PROT_ERR] = "local protection error",
+		[IBV_WC_WR_FLUSH_ERR] = "Work Request Flushed Error",
+		[IBV_WC_MW_BIND_ERR] = "memory management operation error",
+		[IBV_WC_BAD_RESP_ERR] = "bad response error",
+		[IBV_WC_LOC_ACCESS_ERR] = "local access error",
+		[IBV_WC_REM_INV_REQ_ERR] = "remote invalid request error",
+		[IBV_WC_REM_ACCESS_ERR] = "remote access error",
+		[IBV_WC_REM_OP_ERR] = "remote operation error",
+		[IBV_WC_RETRY_EXC_ERR] = "transport retry counter exceeded",
+		[IBV_WC_RNR_RETRY_EXC_ERR] = "RNR retry counter exceeded",
+		[IBV_WC_LOC_RDD_VIOL_ERR] = "local RDD violation error",
+		[IBV_WC_REM_INV_RD_REQ_ERR] = "remote invalid RD request",
+		[IBV_WC_REM_ABORT_ERR] = "aborted error",
+		[IBV_WC_INV_EECN_ERR] = "invalid EE context number",
+		[IBV_WC_INV_EEC_STATE_ERR] = "invalid EE context state",
+		[IBV_WC_FATAL_ERR] = "fatal error",
+		[IBV_WC_RESP_TIMEOUT_ERR] = "response timeout error",
+		[IBV_WC_GENERAL_ERR] = "general error",
+		[IBV_WC_TM_ERR] = "TM error",
+		[IBV_WC_TM_RNDV_INCOMPLETE] = "TM software rendezvous",
+	};
+
+	if (status < IBV_WC_SUCCESS || status > IBV_WC_TM_RNDV_INCOMPLETE)
+		return "unknown";
+
+	return wc_status_str[status];
+}
+
+//
+// infiniband/verbs.h defines ibv_reg_mr as a macro and then uses inline functions to perform some
+// optimization that should result in ibv_reg_mr_iova2 being called on Windows.
+// However, in order to link correctly, there must be a definition of ibv_reg_mr that is not
+// expanded by the macro from verbs.h.
```

Comment:
Could you explain a little bit more? Why the definition is needed if never being used? Is there any place the non-macro version of ibv_reg_mr is referenced?

Reply:
The shorter answer is that there is a call to it, but it is in a code path that will never be executed on Windows because that code path requires __builtin_constant_p to return true. It is a gcc builtin function and I defined it as a macro that always returns 0 for Windows. So, the call is there and requires something to link to, but that code path is dead for all intents and purposes on Windows.

This is a really wonky part of verbs.h. As background, one of my goals was to minimize edits to the header files I pulled from Linux and make the needed adaptations in new code. I could probably have made more substantial edits to verbs.h to avoid this and maybe that's what would be preferable here.

Here's the bit from verbs.h relating to ibv_reg_mr.
```
/**
 * ibv_reg_mr_iova2 - Register memory region with a virtual offset address
 *
 * This version will be called if ibv_reg_mr or ibv_reg_mr_iova were called
 * with at least one potential access flag from the IBV_OPTIONAL_ACCESS_RANGE
 * flags range The optional access flags will be masked if running over kernel
 * that does not support passing them.
 */
struct ibv_mr *ibv_reg_mr_iova2(struct ibv_pd *pd, void *addr, size_t length,
				uint64_t iova, unsigned int access);

/**
 * ibv_reg_mr - Register a memory region
 */
struct ibv_mr *ibv_reg_mr(struct ibv_pd *pd, void *addr, size_t length,
			  int access);
/* use new ibv_reg_mr version only if access flags that require it are used */
__attribute__((__always_inline__)) static inline struct ibv_mr *
__ibv_reg_mr(struct ibv_pd *pd, void *addr, size_t length, unsigned int access,
	     int is_access_const)
{
	if (is_access_const && (access & IBV_ACCESS_OPTIONAL_RANGE) == 0)
		return ibv_reg_mr(pd, addr, length, access);
	else
		return ibv_reg_mr_iova2(pd, addr, length, (uintptr_t)addr,
					access);
}

#define ibv_reg_mr(pd, addr, length, access)                                   \
	__ibv_reg_mr(pd, addr, length, access,                                 \
		     __builtin_constant_p(				       \
			     ((access) & IBV_ACCESS_OPTIONAL_RANGE) == 0))
```
What's your opinion as to what would be best? Better comments in verbs_nd_ibv.c while leaving verbs.h intact or edits to verbs.h to just reduce the whole thing to be a direct call to an ibv_reg_mr implementation and not have to comment about the weirdness.

Reply:
How about having this as the body of `ibv_reg_mr()` instead:
```
	return ibv_reg_mr_iova2(pd, addr, length, (uintptr_t)addr, access);
```
You can then remove those comments.